File: cpp/tests/test_json_output.cpp
#include <gtest/gtest.h>
#include <ccc/json_output.hpp>
#include <cmath>
#include <string>

static const std::string OPENCODE_STDOUT = "{\"response\":\"Hello from OpenCode\"}";

static const std::string CLAUDE_CODE_STDOUT =
    "{\"type\":\"system\",\"subtype\":\"init\",\"session_id\":\"sess-1\",\"tools\":[],\"model\":\"mock\",\"permission_mode\":\"default\"}\n"
    "{\"type\":\"assistant\",\"message\":{\"id\":\"msg_1\",\"type\":\"message\",\"role\":\"assistant\",\"model\":\"mock\","
    "\"content\":[{\"type\":\"text\",\"text\":\"I will help you.\"}],\"stop_reason\":\"end_turn\",\"stop_sequence\":null,"
    "\"usage\":{\"input_tokens\":10,\"output_tokens\":5}},\"session_id\":\"sess-1\"}\n"
    "{\"type\":\"result\",\"subtype\":\"success\",\"cost_usd\":0.003,\"duration_ms\":100,\"duration_api_ms\":80,"
    "\"num_turns\":1,\"result\":\"I will help you.\",\"session_id\":\"sess-1\","
    "\"usage\":{\"input_tokens\":10,\"output_tokens\":5}}\n";

static const std::string CLAUDE_CODE_TOOL_USE =
    "{\"type\":\"system\",\"subtype\":\"init\",\"session_id\":\"sess-2\",\"tools\":[],\"model\":\"mock\",\"permission_mode\":\"default\"}\n"
    "{\"type\":\"tool_use\",\"tool_name\":\"read_file\",\"tool_input\":{\"path\":\"/test.py\"},\"session_id\":\"sess-2\"}\n"
    "{\"type\":\"tool_result\",\"tool_use_id\":\"toolu_1\",\"content\":\"file contents\",\"is_error\":false,\"session_id\":\"sess-2\"}\n"
    "{\"type\":\"result\",\"subtype\":\"success\",\"result\":\"Done\",\"session_id\":\"sess-2\",\"cost_usd\":0.0,\"duration_ms\":200,"
    "\"duration_api_ms\":150,\"num_turns\":2,\"usage\":{\"input_tokens\":20,\"output_tokens\":10}}\n";

static const std::string KIMI_STDOUT = "{\"role\":\"assistant\",\"content\":\"Hello from Kimi\"}";

static const std::string KIMI_WITH_TOOL_CALLS =
    "{\"role\":\"assistant\",\"content\":\"Let me read that.\","
    "\"tool_calls\":[{\"type\":\"function\",\"id\":\"call_1\",\"function\":{\"name\":\"read_file\",\"arguments\":\"{\\\"path\\\":\\\"/src/main.py\\\"}\"}}]}\n"
    "{\"role\":\"tool\",\"content\":[{\"type\":\"text\",\"text\":\"<system>Tool completed successfully.</system>\"},"
    "{\"type\":\"text\",\"text\":\"file contents\"}],\"tool_call_id\":\"call_1\"}\n"
    "{\"role\":\"assistant\",\"content\":\"Done reading the file.\"}";

static const std::string KIMI_THINKING =
    "{\"role\":\"assistant\",\"content\":["
    "{\"type\":\"think\",\"think\":\"I need to analyze this...\",\"encrypted\":null},"
    "{\"type\":\"text\",\"text\":\"Here is my answer\"}]}\n";

TEST(ParseOpenCodeJson, SimpleResponse) {
    auto result = parse_json_output(OPENCODE_STDOUT, "opencode");
    EXPECT_EQ(result.schema_name, "opencode");
    EXPECT_EQ(result.final_text, "Hello from OpenCode");
    EXPECT_EQ(result.events.size(), 1u);
    EXPECT_EQ(result.events[0].event_type, "text");
}

TEST(ParseOpenCodeJson, ErrorResponse) {
    auto result = parse_opencode_json("{\"error\":\"something went wrong\"}");
    EXPECT_EQ(result.error, "something went wrong");
    EXPECT_EQ(result.events.size(), 1u);
    EXPECT_EQ(result.events[0].event_type, "error");
}

TEST(ParseClaudeCodeJson, SimpleResponse) {
    auto result = parse_json_output(CLAUDE_CODE_STDOUT, "claude-code");
    EXPECT_EQ(result.schema_name, "claude-code");
    EXPECT_EQ(result.final_text, "I will help you.");
    EXPECT_EQ(result.session_id, "sess-1");
    EXPECT_NEAR(result.cost_usd, 0.003, 0.0001);
    EXPECT_EQ(result.duration_ms, 100);
}

TEST(ParseClaudeCodeJson, ToolUseAndResult) {
    auto result = parse_claude_code_json(CLAUDE_CODE_TOOL_USE);
    EXPECT_EQ(result.final_text, "Done");
    int tc_count = 0;
    for (auto& e : result.events) if (e.event_type == "tool_use") tc_count++;
    EXPECT_EQ(tc_count, 1);
    for (auto& e : result.events) {
        if (e.event_type == "tool_use") {
            ASSERT_TRUE(e.tool_call.has_value());
            EXPECT_EQ(e.tool_call->name, "read_file");
        }
    }
}

TEST(ParseClaudeCodeJson, ToolResult) {
    auto result = parse_claude_code_json(CLAUDE_CODE_TOOL_USE);
    int tr_count = 0;
    for (auto& e : result.events) if (e.event_type == "tool_result") tr_count++;
    EXPECT_EQ(tr_count, 1);
    for (auto& e : result.events) {
        if (e.event_type == "tool_result") {
            ASSERT_TRUE(e.tool_result.has_value());
            EXPECT_EQ(e.tool_result->tool_call_id, "toolu_1");
            EXPECT_EQ(e.tool_result->content, "file contents");
            EXPECT_FALSE(e.tool_result->is_error);
        }
    }
}

TEST(ParseClaudeCodeJson, ErrorResult) {
    auto result = parse_claude_code_json(
        "{\"type\":\"result\",\"subtype\":\"error\",\"error\":\"rate limited\",\"session_id\":\"s1\"}");
    EXPECT_EQ(result.error, "rate limited");
}

TEST(ParseKimiJson, SimpleResponse) {
    auto result = parse_json_output(KIMI_STDOUT, "kimi");
    EXPECT_EQ(result.schema_name, "kimi");
    EXPECT_EQ(result.final_text, "Hello from Kimi");
}

TEST(ParseKimiJson, ToolCalls) {
    auto result = parse_kimi_json(KIMI_WITH_TOOL_CALLS);
    EXPECT_EQ(result.final_text, "Done reading the file.");
    int tc_count = 0;
    for (auto& e : result.events) if (e.event_type == "tool_call") tc_count++;
    EXPECT_EQ(tc_count, 1);
    for (auto& e : result.events) {
        if (e.event_type == "tool_call") {
            ASSERT_TRUE(e.tool_call.has_value());
            EXPECT_EQ(e.tool_call->name, "read_file");
        }
    }
}

TEST(ParseKimiJson, ToolResult) {
    auto result = parse_kimi_json(KIMI_WITH_TOOL_CALLS);
    int tr_count = 0;
    for (auto& e : result.events) if (e.event_type == "tool_result") tr_count++;
    EXPECT_EQ(tr_count, 1);
    for (auto& e : result.events) {
        if (e.event_type == "tool_result") {
            ASSERT_TRUE(e.tool_result.has_value());
            EXPECT_EQ(e.tool_result->content, "file contents");
        }
    }
}

TEST(ParseKimiJson, Thinking) {
    auto result = parse_kimi_json(KIMI_THINKING);
    int th_count = 0;
    for (auto& e : result.events) if (e.event_type == "thinking") th_count++;
    EXPECT_EQ(th_count, 1);
    for (auto& e : result.events) {
        if (e.event_type == "thinking") {
            EXPECT_EQ(e.thinking, "I need to analyze this...");
        }
    }
    EXPECT_EQ(result.final_text, "Here is my answer");
}

TEST(RenderParsed, OpenCode) {
    auto result = parse_opencode_json(OPENCODE_STDOUT);
    EXPECT_EQ(render_parsed(result), "Hello from OpenCode");
}

TEST(RenderParsed, Claude) {
    auto result = parse_claude_code_json(CLAUDE_CODE_STDOUT);
    auto rendered = render_parsed(result);
    EXPECT_NE(rendered.find("I will help you."), std::string::npos);
}

TEST(RenderParsed, Kimi) {
    auto result = parse_kimi_json(KIMI_STDOUT);
    EXPECT_EQ(render_parsed(result), "Hello from Kimi");
}

TEST(RenderParsed, WithThinking) {
    auto result = parse_kimi_json(KIMI_THINKING);
    auto rendered = render_parsed(result);
    EXPECT_NE(rendered.find("[thinking]"), std::string::npos);
    EXPECT_NE(rendered.find("Here is my answer"), std::string::npos);
}

TEST(RenderParsed, WithToolUse) {
    auto result = parse_claude_code_json(CLAUDE_CODE_TOOL_USE);
    auto rendered = render_parsed(result);
    EXPECT_NE(rendered.find("[tool] read_file"), std::string::npos);
}

TEST(RenderParsed, UnknownSchema) {
    auto result = parse_json_output("{}", "unknown-schema");
    EXPECT_EQ(result.error, "unknown schema: unknown-schema");
}

TEST(RenderParsed, EmptyInput) {
    auto result = parse_opencode_json("");
    EXPECT_EQ(result.final_text, "");
}

TEST(RenderParsed, MalformedJsonSkipped) {
    auto result = parse_opencode_json("not json\n" + OPENCODE_STDOUT);
    EXPECT_EQ(result.final_text, "Hello from OpenCode");
}

File: nim/tests/test_json_output.nim
import std/[unittest, json, options, strutils]
import call_coding_clis/json_output

suite "parseOpencodeJson":
  test "parses response line":
    let r = parseOpencodeJson("""{"response": "hello world"}""")
    check r.schemaName == "opencode"
    check r.finalText == "hello world"
    check r.events.len == 1
    check r.events[0].eventType == "text"
    check r.events[0].text == "hello world"

  test "parses error line":
    let r = parseOpencodeJson("""{"error": "something went wrong"}""")
    check r.error == "something went wrong"
    check r.events.len == 1
    check r.events[0].eventType == "error"

  test "skips invalid JSON lines":
    let r = parseOpencodeJson("not json\n{\"response\": \"ok\"}")
    check r.events.len == 1
    check r.finalText == "ok"

  test "handles multiple lines":
    let input = """{"response": "first"}""" & "\n" & """{"response": "second"}"""
    let r = parseOpencodeJson(input)
    check r.finalText == "second"
    check r.events.len == 2

  test "handles empty input":
    let r = parseOpencodeJson("")
    check r.events.len == 0

suite "parseClaudeCodeJson":
  test "parses system init":
    let r = parseClaudeCodeJson("""{"type": "system", "subtype": "init", "session_id": "abc123"}""")
    check r.sessionId == "abc123"

  test "parses assistant message":
    let r = parseClaudeCodeJson("""{"type": "assistant", "message": {"content": [{"type": "text", "text": "hello"}], "usage": {"input_tokens": 10}}}""")
    check r.finalText == "hello"
    check r.events.len == 1
    check r.events[0].eventType == "assistant"

  test "parses stream text_delta":
    let r = parseClaudeCodeJson("""{"type": "stream_event", "event": {"type": "content_block_delta", "delta": {"type": "text_delta", "text": "hi"}}}""")
    check r.events[0].eventType == "text_delta"
    check r.events[0].text == "hi"

  test "parses stream thinking_delta":
    let r = parseClaudeCodeJson("""{"type": "stream_event", "event": {"type": "content_block_delta", "delta": {"type": "thinking_delta", "thinking": "hmm"}}}""")
    check r.events[0].eventType == "thinking_delta"
    check r.events[0].thinking == "hmm"

  test "parses tool_use":
    let r = parseClaudeCodeJson("""{"type": "tool_use", "tool_name": "read_file", "tool_input": {"path": "/foo"}}""")
    check r.events[0].eventType == "tool_use"
    check r.events[0].toolCall.isSome
    check r.events[0].toolCall.get().name == "read_file"

  test "parses tool_result":
    let r = parseClaudeCodeJson("""{"type": "tool_result", "tool_use_id": "tu1", "content": "file contents", "is_error": false}""")
    check r.events[0].eventType == "tool_result"
    let tr = r.events[0].toolResult.get()
    check tr.toolCallId == "tu1"
    check tr.content == "file contents"
    check tr.isError == false

  test "parses result success":
    let r = parseClaudeCodeJson("""{"type": "result", "subtype": "success", "result": "done", "cost_usd": 0.05, "duration_ms": 1200}""")
    check r.finalText == "done"
    check r.costUsd == 0.05
    check r.durationMs == 1200

  test "parses result error":
    let r = parseClaudeCodeJson("""{"type": "result", "subtype": "error", "error": "fail"}""")
    check r.error == "fail"

  test "parses tool_use_start from stream_event":
    let r = parseClaudeCodeJson("""{"type": "stream_event", "event": {"type": "content_block_start", "content_block": {"type": "tool_use", "id": "tc1", "name": "bash"}}}""")
    check r.events[0].eventType == "tool_use_start"
    check r.events[0].toolCall.get().name == "bash"

suite "parseKimiJson":
  test "parses passthrough TurnBegin":
    let r = parseKimiJson("""{"type": "TurnBegin"}""")
    check r.events[0].eventType == "turnbegin"

  test "parses assistant with string content":
    let r = parseKimiJson("""{"role": "assistant", "content": "hello there"}""")
    check r.finalText == "hello there"
    check r.events[0].eventType == "assistant"

  test "parses assistant with array content":
    let r = parseKimiJson("""{"role": "assistant", "content": [{"type": "text", "text": "result"}, {"type": "think", "think": "reasoning"}]}""")
    check r.events.len == 2
    check r.events[0].eventType == "thinking"
    check r.events[0].thinking == "reasoning"
    check r.events[1].eventType == "assistant"

  test "parses tool calls":
    let r = parseKimiJson("""{"role": "assistant", "content": "", "tool_calls": [{"id": "tc1", "function": {"name": "edit", "arguments": "{}"}}]}""")
    var tcEv: seq[JsonEvent] = @[]
    for ev in r.events:
      if ev.eventType == "tool_call": tcEv.add(ev)
    check tcEv.len == 1
    check tcEv[0].toolCall.get().name == "edit"

  test "parses tool result filtering system tags":
    let r = parseKimiJson("""{"role": "tool", "tool_call_id": "tc1", "content": [{"type": "text", "text": "<system>internal</system>"}, {"type": "text", "text": "visible output"}]}""")
    let tr = r.events[0].toolResult.get()
    check tr.content == "visible output"
    check tr.toolCallId == "tc1"

suite "parseJsonOutput":
  test "dispatches to opencode":
    let r = parseJsonOutput("""{"response": "hi"}""", "opencode")
    check r.schemaName == "opencode"
    check r.finalText == "hi"

  test "dispatches to claude-code":
    let r = parseJsonOutput("""{"type": "system", "subtype": "init", "session_id": "s1"}""", "claude-code")
    check r.sessionId == "s1"

  test "returns error for unknown schema":
    let r = parseJsonOutput("", "unknown")
    doAssert r.error.find("unknown schema") >= 0

suite "renderParsed":
  test "renders text events":
    let output = ParsedJsonOutput(
      schemaName: "opencode",
      events: @[newJsonEvent("text", text = "hello")],
      finalText: "hello"
    )
    check renderParsed(output) == "hello"

  test "renders thinking events":
    let output = ParsedJsonOutput(
      schemaName: "test",
      events: @[newJsonEvent("thinking", thinking = "deep thoughts")]
    )
    check renderParsed(output) == "[thinking] deep thoughts"

  test "renders tool use events":
    let output = ParsedJsonOutput(
      schemaName: "test",
      events: @[newJsonEvent("tool_use", toolCall = some(newToolCall(name = "bash")))]
    )
    check renderParsed(output) == "[tool] bash"

  test "renders error events":
    let output = ParsedJsonOutput(
      schemaName: "test",
      events: @[newJsonEvent("error", text = "oops")]
    )
    check renderParsed(output) == "[error] oops"

  test "falls back to final_text":
    let output = ParsedJsonOutput(
      schemaName: "test",
      events: @[newJsonEvent("system_retry")],
      finalText: "fallback"
    )
    check renderParsed(output) == "fallback"

  test "renders tool_result events":
    let output = ParsedJsonOutput(
      schemaName: "test",
      events: @[newJsonEvent("tool_result", toolResult = some(newToolResult(content = "file data")))]
    )
    check renderParsed(output) == "[tool_result] file data"

  test "renders multiple events with newlines":
    let output = ParsedJsonOutput(
      schemaName: "test",
      events: @[
        newJsonEvent("text", text = "hello"),
        newJsonEvent("tool_use", toolCall = some(newToolCall(name = "bash")))
      ]
    )
    check renderParsed(output) == "hello\n[tool] bash"
